### PR TITLE
Session 5 samarth mayya

### PR DESCRIPTION
**Pull request by SamarthMayya**

### CODE REVIEW

**Comment by SamarthMayya** [comment]
Is my database schema correctly implemented?

File: session_3/blogspot/.browserslistrc [hunk 1]
```diff
@@ -0,0 +1 @@
+defaults
```
Comment (abhishekkumar2718) [inline session_3/blogspot/.browserslistrc:1]: Please don't push session 3 and 4 again

File: session_5/collegedb/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# README
```
Comment (abhishekkumar2718) [inline session_5/collegedb/README.md:1]: Describe the schema and validations in the README

**Review by abhishekkumar2718 — CHANGES_REQUESTED** [review]

File: session_3/blogspot/.browserslistrc (hunk 1 above)
Comment (SamarthMayya) [reply session_3/blogspot/.browserslistrc:1]: How do I do that? I tried fetching upstream, but that didn't change the locally modified files. Should I manually delete the files?

File: session_5/collegedb/db/schema.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#
+# This file is the source Rails uses to define your schema when running `bin/rails
+# db:schema:load`. When creating a new database, `bin/rails db:schema:load` tends to
+# be faster and is potentially less error prone than running all of your
+# migrations from scratch. Old migrations may fail to apply correctly if those
+# migrations use external dependencies or application code.
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2021_02_01_064206) do
```
Comment (abhishekkumar2718) [inline session_5/collegedb/db/schema.rb:13]: Are you not keeping track of submissions made by a student to an assignment?

**Review by abhishekkumar2718 — CHANGES_REQUESTED** [review]

File: session_5/collegedb/db/schema.rb (hunk 3 above)
Comment (SamarthMayya) [reply session_5/collegedb/db/schema.rb:13]: No not yet, I have just implemented the schema... I wanted to make sure that my schema is correct. By the way, to do that I would need an AssignmentsController right?

Comment (abhishekkumar2718) [reply session_5/collegedb/db/schema.rb:13]: Which of the tables would a submission be stored in?

You don't need to write any controller or views for this session.

Comment (SamarthMayya) [reply session_5/collegedb/db/schema.rb:13]: Yeah, but I need a create action to save the article right?

Comment (SamarthMayya) [reply session_5/collegedb/db/schema.rb:13]: I would use has_one_attached :assignment or something in the Assignment model right? And then I would use Active Storage to store the assignment right?

Comment (SamarthMayya) [reply session_5/collegedb/db/schema.rb:13]: Ohh, are you suggesting that I implement a method in the Assignment model itself to save the assignment to the database?

**Comment by SamarthMayya** [comment]
I thought you meant that we can implement this without having any controller or view, and I started racking my brain :). I ended up implementing the controller and view anyway..... Could you check if it is good?

File: session_5/collegedb/db/schema.rb (hunk 3 above)
Comment (abhishekkumar2718) [reply session_5/collegedb/db/schema.rb]: No, what I am suggesting is far simpler - assignments and submissions are _different_.

A course has many assignments and an assignment has many submissions.

Comment (SamarthMayya) [reply session_5/collegedb/db/schema.rb]: Yeah,I got that....

**Comment by SamarthMayya** [comment]
Is there anything else I'm missing on, or is this fine?

**Comment by abhishekkumar2718** [comment]
> Is there anything else I'm missing on, or is this fine?

Can multiple students upload their submission for an assignment? If so, that should be enough

**Comment by SamarthMayya** [comment]
Yeah right? A course has multiple assignments, and each student has a copy of that course. So every student can upload assignments for a course, right?

**Comment by SamarthMayya** [comment]
Suppose I want student 1's submission for assignment 1, in course 1. Then I'd do Student.find(1).courses.find(1).assignments.find(1).submission, right?

**Comment by abhishekkumar2718** [comment]
> Yeah right? A course has multiple assignments, and each student has a copy of that course. So every student can upload assignments for a course, right?


So you are letting students set deadlines and weightage for the assignment? :scream:

Ideally, the faculty must create an assignment to which students upload their solution.

**Comment by SamarthMayya** [comment]
Well, the view only has upload functionality. Are you suggesting I use authorization so that only lecturers can upload assignments and not students?

**Comment by SamarthMayya** [comment]
I haven't even implemented create function in AssignmentController, so there is no way a student could create an assignment. Also, the view for students only allows upload, and not to make any changes to deadline or weightage.

**Comment by SamarthMayya** [comment]
Ok, I realised that multiple students cannot submit assignments.... But why is my schema wrong? A student has multiple courses and each course has many assignments, so doesn't that mean that each student has many assignments? I really am not able to understand what is happening here....

File: session_5/collegedb/app/models/assignment.rb [hunk 4]
```diff
@@ -1,11 +1,11 @@
 class Assignment < ApplicationRecord
   belongs_to :course
-  has_one_attached :submission 
+  has_many :submissions
```
Comment (abhishekkumar2718) [inline session_5/collegedb/app/models/assignment.rb:3]: Should also be `dependent: :destroy`

File: session_5/collegedb/app/models/student.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            errors.add(:roll_number,"should start with year of admission") 
+        end 
+    end 
+
+    def roll_number_should_contain_course_code
+        code = branch.split.map(&:first).join
+        if code.starts_with?('CS') && !roll_number.include?('CS') 
+            errors.add(:roll_number,"should contain course code")
+        elsif code.starts_with?('ME') 
+            if branch = 'Mechanical Engineering' && !roll_number.include?('ME')
```
Comment (abhishekkumar2718) [inline session_5/collegedb/app/models/student.rb:28]: Woah, that's wrong - should be `==` and not `=`

**Review by abhishekkumar2718 — CHANGES_REQUESTED** [review]

**Comment by SamarthMayya** [comment]
Anything else I should do to improve the model? This session is exclusively model related right?

**Comment by abhishekkumar2718** [comment]
> Anything else I should do to improve the model? This session is exclusively model related right?

Yeah, it is - if you feel there is something to learnt in implementing controllers and views, go for it. 

Alternatively, you can take a look at the [What's Next](https://github.com/IRIS-NITK/IRIS-RoR-Bootcamp-2020#whats-next) section in the README to learn more advanced Rails topics

**Pull request closed without merging**